# Neural network using torch

## Importing required libraries

In [1]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable

## Training settings

In [2]:
batch_size = 64

## Getting MNIST dataset

In [3]:
train_dataset = datasets.MNIST(root='./mnist_data/',
                               train=True,
                               transform=transforms.ToTensor(),
                               download=True)

test_dataset = datasets.MNIST(root='./mnist_data/',
                              train=False,
                              transform=transforms.ToTensor())

Processing...
Done!


## Data Loader (Input Pipeline)

In [4]:
train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=batch_size,
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                          batch_size=batch_size,
                                          shuffle=False)

## 1-Setting neural network model

In [5]:
class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.l1 = nn.Linear(784, 520) # input x is 784 so we start from here
        self.l2 = nn.Linear(520, 320)
        self.l3 = nn.Linear(320, 240)
        self.l4 = nn.Linear(240, 120)
        self.l5 = nn.Linear(120, 10) # ouput is of ten categories in number. So we end it with 10

    def forward(self, x):
        x = x.view(-1, 784)  # Flatten the data (n, 1, 28, 28)-> (n, 784)
        x = F.relu(self.l1(x))
        x = F.relu(self.l2(x))
        x = F.relu(self.l3(x))
        x = F.relu(self.l4(x))
        return self.l5(x)

In [6]:
model = Net()

## 2-Construct loss and optimizer

In [7]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

## 3-Training Cycle
 - Forward
 - backward
 - update

In [8]:
def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = Variable(data), Variable(target)
        optimizer.zero_grad()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))


## 4- Testing the model

In [11]:
def test():
    model.eval()
    test_loss = 0
    correct = 0
    for data, target in test_loader:
        data, target = Variable(data, volatile=True), Variable(target)
        output = model(data)
        # sum up batch loss
        test_loss += criterion(output, target).item()
        # get the index of the max
        pred = output.data.max(1, keepdim=True)[1]
        correct += pred.eq(target.data.view_as(pred)).cpu().sum()

    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


In [12]:
#Full range of train and test set where we can see how loss function of our prediction has changed
for epoch in range(1, 10):
    train(epoch)
    test()

Train Epoch: 1 [0/60000 (0%)]	Loss: 1.465507
Train Epoch: 1 [640/60000 (1%)]	Loss: 1.308517
Train Epoch: 1 [1280/60000 (2%)]	Loss: 1.364752
Train Epoch: 1 [1920/60000 (3%)]	Loss: 1.413621
Train Epoch: 1 [2560/60000 (4%)]	Loss: 1.342042
Train Epoch: 1 [3200/60000 (5%)]	Loss: 1.257293
Train Epoch: 1 [3840/60000 (6%)]	Loss: 1.298469
Train Epoch: 1 [4480/60000 (7%)]	Loss: 1.291172
Train Epoch: 1 [5120/60000 (9%)]	Loss: 0.979086
Train Epoch: 1 [5760/60000 (10%)]	Loss: 1.283813
Train Epoch: 1 [6400/60000 (11%)]	Loss: 1.154893
Train Epoch: 1 [7040/60000 (12%)]	Loss: 0.998969
Train Epoch: 1 [7680/60000 (13%)]	Loss: 0.832424
Train Epoch: 1 [8320/60000 (14%)]	Loss: 1.016709
Train Epoch: 1 [8960/60000 (15%)]	Loss: 1.229345
Train Epoch: 1 [9600/60000 (16%)]	Loss: 0.798754
Train Epoch: 1 [10240/60000 (17%)]	Loss: 0.993614
Train Epoch: 1 [10880/60000 (18%)]	Loss: 0.854521
Train Epoch: 1 [11520/60000 (19%)]	Loss: 0.828913
Train Epoch: 1 [12160/60000 (20%)]	Loss: 1.000607
Train Epoch: 1 [12800/60000 (

C:\Users\Hassan\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: UserWarning: volatile was removed and now has no effect. Use `with torch.no_grad():` instead.
  



Test set: Average loss: 0.0066, Accuracy: 8726/10000 (87%)

Train Epoch: 2 [0/60000 (0%)]	Loss: 0.375830
Train Epoch: 2 [640/60000 (1%)]	Loss: 0.478374
Train Epoch: 2 [1280/60000 (2%)]	Loss: 0.400222
Train Epoch: 2 [1920/60000 (3%)]	Loss: 0.330684
Train Epoch: 2 [2560/60000 (4%)]	Loss: 0.533795
Train Epoch: 2 [3200/60000 (5%)]	Loss: 0.627898
Train Epoch: 2 [3840/60000 (6%)]	Loss: 0.389320
Train Epoch: 2 [4480/60000 (7%)]	Loss: 0.320679
Train Epoch: 2 [5120/60000 (9%)]	Loss: 0.457995
Train Epoch: 2 [5760/60000 (10%)]	Loss: 0.483714
Train Epoch: 2 [6400/60000 (11%)]	Loss: 0.360923
Train Epoch: 2 [7040/60000 (12%)]	Loss: 0.617252
Train Epoch: 2 [7680/60000 (13%)]	Loss: 0.245925
Train Epoch: 2 [8320/60000 (14%)]	Loss: 0.678559
Train Epoch: 2 [8960/60000 (15%)]	Loss: 0.395057
Train Epoch: 2 [9600/60000 (16%)]	Loss: 0.496613
Train Epoch: 2 [10240/60000 (17%)]	Loss: 0.344263
Train Epoch: 2 [10880/60000 (18%)]	Loss: 0.414041
Train Epoch: 2 [11520/60000 (19%)]	Loss: 0.350224
Train Epoch: 2 [12

Train Epoch: 3 [43520/60000 (72%)]	Loss: 0.242689
Train Epoch: 3 [44160/60000 (74%)]	Loss: 0.159345
Train Epoch: 3 [44800/60000 (75%)]	Loss: 0.173540
Train Epoch: 3 [45440/60000 (76%)]	Loss: 0.236224
Train Epoch: 3 [46080/60000 (77%)]	Loss: 0.520841
Train Epoch: 3 [46720/60000 (78%)]	Loss: 0.349474
Train Epoch: 3 [47360/60000 (79%)]	Loss: 0.228130
Train Epoch: 3 [48000/60000 (80%)]	Loss: 0.344371
Train Epoch: 3 [48640/60000 (81%)]	Loss: 0.196172
Train Epoch: 3 [49280/60000 (82%)]	Loss: 0.233393
Train Epoch: 3 [49920/60000 (83%)]	Loss: 0.205863
Train Epoch: 3 [50560/60000 (84%)]	Loss: 0.393267
Train Epoch: 3 [51200/60000 (85%)]	Loss: 0.210180
Train Epoch: 3 [51840/60000 (86%)]	Loss: 0.367082
Train Epoch: 3 [52480/60000 (87%)]	Loss: 0.274039
Train Epoch: 3 [53120/60000 (88%)]	Loss: 0.173376
Train Epoch: 3 [53760/60000 (90%)]	Loss: 0.201184
Train Epoch: 3 [54400/60000 (91%)]	Loss: 0.284615
Train Epoch: 3 [55040/60000 (92%)]	Loss: 0.262176
Train Epoch: 3 [55680/60000 (93%)]	Loss: 0.127078


Train Epoch: 5 [26880/60000 (45%)]	Loss: 0.108311
Train Epoch: 5 [27520/60000 (46%)]	Loss: 0.191536
Train Epoch: 5 [28160/60000 (47%)]	Loss: 0.066319
Train Epoch: 5 [28800/60000 (48%)]	Loss: 0.096661
Train Epoch: 5 [29440/60000 (49%)]	Loss: 0.143405
Train Epoch: 5 [30080/60000 (50%)]	Loss: 0.100283
Train Epoch: 5 [30720/60000 (51%)]	Loss: 0.096482
Train Epoch: 5 [31360/60000 (52%)]	Loss: 0.235376
Train Epoch: 5 [32000/60000 (53%)]	Loss: 0.048751
Train Epoch: 5 [32640/60000 (54%)]	Loss: 0.085049
Train Epoch: 5 [33280/60000 (55%)]	Loss: 0.120577
Train Epoch: 5 [33920/60000 (57%)]	Loss: 0.040202
Train Epoch: 5 [34560/60000 (58%)]	Loss: 0.217670
Train Epoch: 5 [35200/60000 (59%)]	Loss: 0.094831
Train Epoch: 5 [35840/60000 (60%)]	Loss: 0.238977
Train Epoch: 5 [36480/60000 (61%)]	Loss: 0.178521
Train Epoch: 5 [37120/60000 (62%)]	Loss: 0.132406
Train Epoch: 5 [37760/60000 (63%)]	Loss: 0.056131
Train Epoch: 5 [38400/60000 (64%)]	Loss: 0.082624
Train Epoch: 5 [39040/60000 (65%)]	Loss: 0.108985


Train Epoch: 7 [10240/60000 (17%)]	Loss: 0.193060
Train Epoch: 7 [10880/60000 (18%)]	Loss: 0.153921
Train Epoch: 7 [11520/60000 (19%)]	Loss: 0.022341
Train Epoch: 7 [12160/60000 (20%)]	Loss: 0.096721
Train Epoch: 7 [12800/60000 (21%)]	Loss: 0.039802
Train Epoch: 7 [13440/60000 (22%)]	Loss: 0.350112
Train Epoch: 7 [14080/60000 (23%)]	Loss: 0.016778
Train Epoch: 7 [14720/60000 (25%)]	Loss: 0.096543
Train Epoch: 7 [15360/60000 (26%)]	Loss: 0.124670
Train Epoch: 7 [16000/60000 (27%)]	Loss: 0.183723
Train Epoch: 7 [16640/60000 (28%)]	Loss: 0.156859
Train Epoch: 7 [17280/60000 (29%)]	Loss: 0.047206
Train Epoch: 7 [17920/60000 (30%)]	Loss: 0.053374
Train Epoch: 7 [18560/60000 (31%)]	Loss: 0.103679
Train Epoch: 7 [19200/60000 (32%)]	Loss: 0.067131
Train Epoch: 7 [19840/60000 (33%)]	Loss: 0.063463
Train Epoch: 7 [20480/60000 (34%)]	Loss: 0.113145
Train Epoch: 7 [21120/60000 (35%)]	Loss: 0.023035
Train Epoch: 7 [21760/60000 (36%)]	Loss: 0.054133
Train Epoch: 7 [22400/60000 (37%)]	Loss: 0.102715


Train Epoch: 8 [54400/60000 (91%)]	Loss: 0.088805
Train Epoch: 8 [55040/60000 (92%)]	Loss: 0.140128
Train Epoch: 8 [55680/60000 (93%)]	Loss: 0.050760
Train Epoch: 8 [56320/60000 (94%)]	Loss: 0.122722
Train Epoch: 8 [56960/60000 (95%)]	Loss: 0.021657
Train Epoch: 8 [57600/60000 (96%)]	Loss: 0.109530
Train Epoch: 8 [58240/60000 (97%)]	Loss: 0.100695
Train Epoch: 8 [58880/60000 (98%)]	Loss: 0.175598
Train Epoch: 8 [59520/60000 (99%)]	Loss: 0.021141

Test set: Average loss: 0.0017, Accuracy: 9686/10000 (96%)

Train Epoch: 9 [0/60000 (0%)]	Loss: 0.035383
Train Epoch: 9 [640/60000 (1%)]	Loss: 0.088763
Train Epoch: 9 [1280/60000 (2%)]	Loss: 0.085770
Train Epoch: 9 [1920/60000 (3%)]	Loss: 0.075744
Train Epoch: 9 [2560/60000 (4%)]	Loss: 0.038871
Train Epoch: 9 [3200/60000 (5%)]	Loss: 0.040269
Train Epoch: 9 [3840/60000 (6%)]	Loss: 0.103701
Train Epoch: 9 [4480/60000 (7%)]	Loss: 0.050522
Train Epoch: 9 [5120/60000 (9%)]	Loss: 0.072467
Train Epoch: 9 [5760/60000 (10%)]	Loss: 0.050432
Train Epoch:

**At final setp, we manage to have a 97% accuracy and 0.0015 average loss shown from our test set**